src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/WCFUpdateStep.cs

```diff
@@ -0,0 +1,226 @@
+            _packageUpdater = UpdaterFactory.GetPackageUpdater(_path["proj"].First(), Logger);
+            if (_path.ContainsKey("directives"))
+            {
+                _directiveUpdaters = UpdaterFactory.GetDirectiveUpdaters(_path["directives"], Logger);
+            }
+
+            if (_configUpdater is not null)
+            {
+                var configContext = UpdateRunner.GetContext(_configUpdater);
+                _sourceCodeUpdater = UpdaterFactory.GetSourceCodeUpdater(_path["main"].First(), configContext, Logger);
+            }
+
+            if (_configUpdater is null || _packageUpdater is null || _sourceCodeUpdater is null || (_path.ContainsKey("directives") && _directiveUpdaters is null))
+            {
+                Logger.LogWarning("Unexpected error happened when trying to construct the updaters. Please review error message and log.");
+                return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Failed, "Updaters cannot be constructed. Please review error message and log for more information.", BuildBreakRisk.Medium));
+            }
+
+            return Task.FromResult(new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, "WCF updaters are initialized. Ready to update.", BuildBreakRisk.None));
+        }
+
+        public Task<UpgradeStepApplyResult> Apply()
+        {
+            // Run updates
+            var projFile = UpdateRunner.PackageUpdate(_packageUpdater!, Logger);
+            var config = UpdateRunner.ConfigUpdate(_configUpdater!, Logger);
+            var source = UpdateRunner.SourceCodeUpdate(_sourceCodeUpdater!, Logger);
+            var directives = new List<SyntaxNode>();
+            if (_directiveUpdaters is not null)
+            {
+                directives = UpdateRunner.DirectiveUpdate(_directiveUpdaters, Logger);
+            }
+
+            // check for null changes
+            if (projFile is null || config is null || source is null || directives is null)
+            {
+                Logger.LogWarning("Unexpected error happened when trying to making changes to original files. Please review error message and log.");
+                return Task.FromResult(new UpgradeStepApplyResult(UpgradeStepStatus.Failed, "Project failed to port to CoreWCF. Please review error message and log for more information."));
+            }
+
+            // Write updates
+            UpdateRunner.WritePackageUpdate(projFile, _path!["proj"].First(), Logger);
+            UpdateRunner.WriteConfigUpdate(config[0], config[1], _path["config"].First(), Logger);
+            UpdateRunner.WriteSourceCodeUpdate(source, _path["main"].First(), Logger);
+            if (directives.Count > 0)
+            {
+                UpdateRunner.WriteDirectiveUpdate(directives, _path["directives"], Logger);
+            }
+
+            Logger.LogInformation("Project was successfully updated to use CoreWCF services. Please review changes.");
```

Comment:
Since the project will have changed directly on disk and not via the project interfaces, it's a best practice to reload the project workspace. That way, we can be sure that future upgrade steps will see your changes reflected in what's loaded. You can do that with `IUpgradeContext.ReloadWorkspaceAsync`.